### PR TITLE
memcached support

### PR DESCRIPTION
**Pull request by urso**
memcached support for text based protocol tcp/udp. Binary memcache protocol will follow soon.

Please do not merge until binary protocol + tests are included.

Memcached support requires libbeat feature/protocol-helpers1 for parsing.

### CODE REVIEW

**Comment by urso** [comment]
build fails due to libbeat feature/protocol-helpers1 pull request missing yet.

File: protos/memcache/plugin.go [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+}
+
+func (mc *Memcache) ParseUdp(pkt *protos.Packet) {
+	defer logp.Recover("ParseMemcache(UDP) exception")
+
+	buffer := streambuf.NewFixed(pkt.Payload)
+
+	requestId, _ := buffer.NetUint16()
+	seqNumber, _ := buffer.NetUint16()
+	numDatagrams, _ := buffer.NetUint16()
```
Comment (urso) [inline protos/memcache/plugin.go]: ups, 0 check missing

File: protos/memcache/plugin.go [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+		if udpMsg.count == udpMsg.numDatagrams {
+			parsePacket = true
+			buffer = streambuf.New(nil)
+			for _, payload := range udpMsg.datagrams {
+				buffer.Append(payload)
+			}
+			udpMsg.datagrams = nil
+			parsePacket = true
+		}
+	}
```
Comment (urso) [inline protos/memcache/plugin.go]: TODO: add mode for ignoring the data part (up to 1MB) by trying to parse first datargram + continue until command was fully read.

File: protos/memcache/plugin.go [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+		results,
+	)
+}
+
+// Called to return the configured ports
+func (mc *Memcache) GetPorts() []int {
+	return mc.Ports.Ports
+}
+
+func (mc *Memcache) ParseUdp(pkt *protos.Packet) {
```
Comment (urso) [inline protos/memcache/plugin.go]: TODO: function too big, create some helpers

**Comment by tsg** [comment]
For the Travis build to work, you probably need to do to: `godep update ...` and commit the changes. Check that the new packages are actually added to the `_workspace` folder.

**Comment by urso** [comment]
this does the trick, thanks.

**Comment by tsg** [comment]
Had a look again. LGTM!

The only comment I have is that maybe it would make sense to use git rebase to cleanup the commit history a bit by squashing some commits together. I think the Elasticsearch team squashes all pull requests in a single commit. I'd say we don't have to follow that rule, but still should try to keep a clean history when possible.

**Comment by ruflin** [comment]
I'm actually not sure what would happen if @urso rebases now, as it is already pushed to Github. What I do quite frequently as with rebase I often end up with lots of merge, I do a lot of small local commits (before I push) and the use git reset --soft somecommit and do one big commit.

**Comment by tsg** [comment]
He would `git push -f` to github :-). But since that's in his private repo, it's ok.

File: protos/memcache/binary_test.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
+package memcache
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	// "github.com/elastic/libbeat/common/streambuf"
```
Comment (ruflin) [inline protos/memcache/binary_test.go]: Should be remove if not used anymore

File: protos/memcache/commands.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
+package memcache
+
+import (
+	// "bytes"
```
Comment (ruflin) [inline protos/memcache/commands.go]: Remove

File: protos/memcache/errors.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
+package memcache
+
+import (
+	"errors"
+)
+
+// error definitions
+
+// TODO: move some to libbeat?
```
Comment (ruflin) [inline protos/memcache/errors.go]: I would more see it generalised in packetbeat as libbeat should not know about ports etc, that is specific to packetbeat.

File: protos/memcache/memcache.go [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+	data []byte
+}
+
+type memcacheStat struct {
+	Name  memcacheString `json:"name"`
+	Value memcacheString `json:"value"`
+}
+
+// Ports configuration
+// TODO: move to libbeat?
```
Comment (ruflin) [inline protos/memcache/memcache.go]: See comment above

File: protos/memcache/memcache.go [hunk 8]
```diff
@@ -0,0 +1,506 @@
...
+type PortsConfig struct {
+	Ports []int
+}
+
+// debug logging support.
+
+var debug = makeDebug("memcache")
+
+func makeDebug(module string) func(string, ...interface{}) {
```
Comment (ruflin) [inline protos/memcache/memcache.go]: Is this the same as MakeDebug in logp? If yes, I suggest to use the libbeat one.

File: protos/memcache/memcache.go [hunk 9]
```diff
@@ -0,0 +1,506 @@
...
+func (mc memcacheString) MarshalJSON() ([]byte, error) {
+	s := string(mc.raw)
+	return json.Marshal(s)
+	/*
+		buf := bytes.NewBuffer(nil)
+		buf.Grow(len(mc.raw) + 2)
+		buf.WriteByte('"')
+		buf.Write(mc.raw) // TODO: escape string
+		buf.WriteByte('"')
+		return buf.Bytes(), nil
```
Comment (ruflin) [inline protos/memcache/memcache.go]: Remove if not needed anymore

File: protos/memcache/parse.go [hunk 10]
```diff
@@ -0,0 +1,177 @@
+package memcache
+
+import (
+	// "runtime"
```
Comment (ruflin) [inline protos/memcache/parse.go]: Remove

File: protos/memcache/parse.go [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+func (p *parser) yieldNoData(buf *streambuf.Buffer) parseResult {
+	return p.yield(buf.BufferConsumed())
+}
+
+func (p *parser) failing(err error) parseResult {
+	// buf := make([]byte, 1024)
+	// runtime.Stack(buf, false)
+	// debug("failing, stacktrace: %s", buf)
```
Comment (ruflin) [inline protos/memcache/parse.go]: Remove

File: protos/memcache/text.go [hunk 12]
```diff
@@ -0,0 +1,744 @@
+package memcache
+
+import (
+	"bytes"
+	"fmt"
+	// "runtime"
+	// "time"
```
Comment (ruflin) [inline protos/memcache/text.go]: Remove

File: protos/memcache/text_test.go [hunk 13]
```diff
@@ -0,0 +1,221 @@
+package memcache
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	// "github.com/elastic/libbeat/common/streambuf"
```
Comment (ruflin) [inline protos/memcache/text_test.go]: Remove

File: protos/memcache/text.go [hunk 14]
```diff
@@ -0,0 +1,744 @@
+package memcache
+
```
Comment (ruflin) [inline protos/memcache/text.go]: Having a brief description/documentation what this file is about would be nice. It was not obvious to me at first.

File: tests/gen/memcache/gen_all.sh [hunk 15]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
```
Comment (ruflin) [inline tests/gen/memcache/gen_all.sh]: Isn't it necessary to update the test Makefile to run this script?

**Comment by ruflin** [comment]
I added a few comments which are mostly cosmetic. I like that some parts of the code are really well documented which makes it much easier to read. 

Make sure to run `gofmt -w .` as I think it will cleanup some more stuff.

Besides the comments in the code, this looks really good and from my side is ready to merge.

File: tests/gen/memcache/gen_all.sh (hunk 15 above)
Comment (urso) [inline tests/gen/memcache/gen_all.sh]: No, don run this every time. This script is about generating the different scenarios:

Reasons not to run every single time:
- you need a memcache instance to create pcaps => commit generated pcaps
- You need privileges to sniff traffic via tcpdump
- Also running every single time would make tests unstable as UDP can drop packets (even between local VMs as I've found)

Will add a README explaining how to use it

File: protos/memcache/text.go (hunk 14 above)
Comment (urso) [inline protos/memcache/text.go]: good point, will add some short description to every file

**Comment by urso** [comment]
my editor is already running 'gofmt -w .' on every save ;)

File: tests/gen/memcache/gen_all.sh (hunk 15 above)
Comment (ruflin) [inline tests/gen/memcache/gen_all.sh]: Ok, got it. So the output of this is also checked into the repo.

**Pull request merged**